fix: Issue #9364 JSON config validation

gsergey418alt opened the pull request:
Fixes #9364. Disallow unknown fields when updating the config with `ipfs config`
Before:
```
$ ipfs config --json -- Provider  '{"Foo":""}' &&  ipfs config --json -- Provider && echo ok
{
  "Foo": ""
}
ok
$
```
After:
```
$ ipfs config --json -- Provider  '{"Foo":""}' &&  ipfs config --json -- Provider && echo ok
Error: failed to set config value: failure to decode config: json: unknown field "Foo" (maybe use --json?)
$
```

<!--
Please update docs/changelogs/ if you're modifying Go files. If your change does not require a changelog entry, please do one of the following:
- add `[skip changelog]` to the PR title
- label the PR with `skip/changelog`
-->

guillaumemichel commented:
Triage notes:
* Changelog should be moved to 0.34
* Strict config file parsing should be optional (CLI flag e.g `ipfs daemon --disallow-unknown-config-fields`, `false` by default) to avoid breaking existing users

gsergey418alt commented:
Hello, I don't see how this could break backwards compatibility with other versions, since invalid values aren't read by the config anyway and I don't see any other reason for users to set invalid config fields. The FromMap method is only called in the SetConfigKey method, which is used in gc.go once for initializing default values and from the CLI e.g. `ipfs config`. Furthermore the comment in SetConfigKey in fsrepo.go hints that this was the intended functionality that just wasn't implemented.
```

	// This step doubles as to validate the map against the struct
	// before serialization
	conf, err := config.FromMap(mapconf)
	if err != nil {
		return err
	}

```

gsergey418alt commented:
Yeah, you're right, that would break existing users, will fix that tomorrow.

guillaumemichel commented on config/config.go:
```diff
@@ -152,3 +176,42 @@ func (c *Config) Clone() (*Config, error) {
 }
+
+// Check if the provided key is present in the structure.
+func CheckKey(key string) error {
```
Why not use `decoder.DisallowUnknownFields()` like you initially suggested? The code change would be much smaller.

gsergey418alt replied:
Because that would disallow unknown fields for the entire config, not just the new added entries, which would possibly break existing users. Furthermore, this would only work when running `ipfs config`, and I think it's pretty illogical to require the config to be strict only when setting new options. Also, it's probably not the functionality requested in the issue #9364.

guillaumemichel replied:
Makes sense, thanks for the explanation

guillaumemichel reviewed (changes requested):
Thanks @gsergey418alt for your contribution.

Could you add a flag to enable/disable strict config JSON parsing? e.g `ipfs daemon --allow-unknown-config-fields`, `true` by default.

This would allow existing users to pass that flag to continue using the same config, and would facilitate development when comparing different versions with the same config.

gsergey418alt commented on test/sharness/t0021-config.sh:
```diff
@@ -13,41 +13,17 @@ test_config_cmd_set() {
+    ipfs config $cfg_flags \"$cfg_key\" \"$cfg_val\"
+  "
 
-  # also test our lib function. it should work too.
-  cfg_key="Lib.$cfg_key"
-  test_expect_success "test_config_set succeeds" '
-    test_config_set $cfg_flags "$cfg_key" "$cfg_val"
-  '
-
-  test_expect_success "test_config_set value looks good" '
-    echo "$cfg_val" >expected &&
-    ipfs config "$cfg_key" >actual &&
```
This just adds the same config options in the "Lib" subtree of the config. Correct me if I'm wrong, but this doesn't seem to be parsed or used anywhere in the codebase and is just as good as adding other arbitrary options.

gsergey418alt commented:
Docker commands still failing in CI, maybe some issue on your side? I don't see anything related to `ipfs config`.

gsergey418alt commented:
@guillaumemichel The existing users can keep using the same config, the validation is only done on the `ipfs config` command.

guillaumemichel commented on config/config.go:
```diff
@@ -137,6 +138,29 @@ func ToMap(conf *Config) (map[string]interface{}, error) {
 	return m, nil
 }
 
+// Convert config to a map, without using encoding/json.
+func ReflectToMap(conf interface{}) map[string]interface{} {
```
what is the reason to define a new function, and not use `ToMap` defined above?

gsergey418alt replied:
There's no way to get encoding/json (which is used in the ToMap function) to ignore the struct tags like `json:",omitempty"` that hide the map fields. That means that the user wouldn't be able to set these fields, since this map is used to validate user's input key from `ipfs config`. Had to use the `reflect` package to make that happen.

gsergey418alt commented:
Same stuff with the docker commands

guillaumemichel commented:
I just checked locally, and only this PR is failing the docker sharness tests. `master` doesn't have this issue. 

* [this CID](https://github.com/gsergey418alt/kubo/blob/ab2657ead7307c4673f54e1a5a25651bf3efdaea/test/sharness/t0002-docker-image.sh#L53) probably needs to be adjusted. Not sure what it corresponds to, but maybe one of the [files defined above](https://github.com/gsergey418alt/kubo/blob/ab2657ead7307c4673f54e1a5a25651bf3efdaea/test/sharness/t0002-docker-image.sh#L39-L40)?
* [this test](https://github.com/gsergey418alt/kubo/blob/ab2657ead7307c4673f54e1a5a25651bf3efdaea/test/sharness/t0002-docker-image.sh#L66-L72) seems directly linked with the config

gsergey418alt commented:
@guillaumemichel Yep, seen that test. I fixed the values to be valid, but it's still failing because the docker container isn't starting up.

gsergey418alt commented:
Maybe open a new PR?

gsergey418alt commented:
Maybe open a new PR?

guillaumemichel commented:
Turns out `Datastore.Path` isn't a valid config anymore: `old style datatstore config detected`, which is why the docker tests were failing. I'll open a new PR to remove it from the `ipfs config` help.

gsergey418alt commented:
Awesome, good catch.

guillaumemichel commented:
I updated the `ReflectToMap` since it was missing some cases, and added a test for it.

@gsergey418alt could you please adapt `CheckKey` accordingly, and add extensive sharness tests e.g in `t0021-config.sh` covering all possible key types (e.g bool, int, string, list, nested struct, etc.) both in json and "." (`Key1.Key2.Key3`) format where possible.

gsergey418alt commented:
Good work, but you would need a way to distinguish between in-config maps and structure. In my implementation, this was done via the "map" value of a field, so when the validator encounters this field, it knows that the user is trying to set nested value inside a map and it discards the rest of key. I think I should've named the function something like `GetValidationMap` or something

guillaumemichel commented:
> you would need a way to distinguish between in-config maps and structure

 could you expand on that? I am not sure to understand the difference in practice.

gsergey418alt commented:
If we run `ipfs config --json Routing.Routers.Test "{\"Parameters\":\"Test\",\"Type\":\"Test\"}"`, we need a way to know that `.Test` is a user-defined entry in the `Routers` map and isn't supposed to be validated.

guillaumemichel commented:
> @guillaumemichel If we run ipfs config --json Routing.Routers.Test "{\"Parameters\":\"Test\",\"Type\":\"Test\"}", we need a way to know that .Test is a user-defined entry in the Routers map and isn't supposed to be validated.

Ok I see. Would that work for maps that have nested types?

gsergey418alt commented:
@guillaumemichel Hmm, haven't thought about that actually. Maybe we could create a "ref" element inside that map that could be validated against if you want that covered.

gsergey418alt commented:
@guillaumemichel I've added sharness test for other data types, changed function name to `GetValidationMap`. I've also changed the behavior of the function for maps, as it was in my previous implementation, without validation for map-nested structures. I won't be active until Monday, so have a good weekend!

gsergey418alt commented:
@guillaumemichel I've solved the problem with validating structs inside maps, and it now works as expected. Added test for PublicGateways.

guillaumemichel commented on config/config.go:
```diff
@@ -137,6 +138,71 @@ func ToMap(conf *Config) (map[string]interface{}, error) {
 }
 
+// Convert config to a map, without using encoding/json, since
+// zero/empty/'omitempty' fields are exclused by encoding/json during
```
```suggestion
// zero/empty/'omitempty' fields are excluded by encoding/json during
```

guillaumemichel commented:
> @guillaumemichel I've solved the problem with validating structs inside maps, and it now works as expected. Added test for PublicGateways.

Looks good thanks!

A [sharness test](https://github.com/ipfs/kubo/blob/4bd79bdbcaa5401c5b02cdabc35f26f310717b5b/test/sharness/t0280-plugin-peerlog.sh#L32) is still failing, probably due to the following config command:
```
ipfs config --json Plugins.Plugins.peerlog.Config.Enabled true
```

Additionally, could you also add a test `TestCheckKey` in [`config_test.go`](https://github.com/ipfs/kubo/blob/master/config/config_test.go) testing a few possible keys?

gsergey418alt commented:
Added handling for interface{} in the config, like Plugins.Plugins.peerlog.Config.Enabled

guillaumemichel commented on config/config.go:
```diff
@@ -152,3 +218,38 @@ func (c *Config) Clone() (*Config, error) {
+	for i, part := range parts {
+		mapCursor, ok = cursor.(map[string]interface{})
+		if !ok {
+			if v, ok := cursor.(string); ok && v == "any" {
+				return nil
+			}
```
@gsergey418alt would that work? Using `"any"` as key seems strange.

```suggestion
			if cursor == nil {
				return nil
			}
```

gsergey418alt replied:
There are other empty strings in the output of ReflectToMap. Nil pointer could work, but "any" is perhaps more expressive e. g. this key contain anything, stop validating. I can change it to nil ptr if you want to.

gsergey418alt replied:
NVM, It wouldn't work with empty strings, but it would still fail at a later stage.

gsergey418alt replied:
Changed "any" to nil

guillaumemichel reviewed (approved):
LGTM, thanks for your contribution @gsergey418alt!
